exchanges/gateio/gateio_websocket.go

```diff
@@ -38,697 +66,832 @@ func (g *Gateio) WsConnect() error {
+			Date:           resp.Result[x].CreateTimeMs,
+			LastUpdated:    resp.Result[x].UpdateTimeMs,
 		}
+	}
+	return nil
+}
 
-		if IsSnapshot {
-			var newOrderBook orderbook.Base
-			newOrderBook.Asks = asks
-			newOrderBook.Bids = bids
-			newOrderBook.Asset = asset.Spot
-			newOrderBook.Pair = p
-			newOrderBook.Exchange = g.Name
-			newOrderBook.VerifyOrderbook = g.CanVerifyOrderbook
-
-			err = g.Websocket.Orderbook.LoadSnapshot(&newOrderBook)
-			if err != nil {
-				return err
-			}
-		} else {
-			err = g.Websocket.Orderbook.Update(&orderbook.Update{
-				Asks:       asks,
-				Bids:       bids,
-				Pair:       p,
-				UpdateTime: time.Now(),
-				Asset:      asset.Spot,
-			})
-			if err != nil {
-				return err
-			}
-		}
-	case strings.Contains(result.Method, "kline"):
-		var data []interface{}
-		err = json.Unmarshal(result.Params[0], &data)
-		if err != nil {
-			return err
-		}
-		open, err := strconv.ParseFloat(data[1].(string), 64)
-		if err != nil {
-			return err
-		}
-		closePrice, err := strconv.ParseFloat(data[2].(string), 64)
-		if err != nil {
-			return err
-		}
-		high, err := strconv.ParseFloat(data[3].(string), 64)
+func (g *Gateio) processUserPersonalTrades(data []byte) error {
+	type response struct {
+		Time    int64                 `json:"time"`
+		Channel string                `json:"channel"`
+		Event   string                `json:"event"`
+		Result  []WsUserPersonalTrade `json:"result"`
+	}
+	var resp response
+	err := json.Unmarshal(data, &resp)
+	if err != nil {
+		return err
+	}
+	fills := make([]fill.Data, len(resp.Result))
+	for x := range fills {
+		currencyPair, err := currency.NewPairFromString(resp.Result[x].CurrencyPair)
 		if err != nil {
 			return err
 		}
-		low, err := strconv.ParseFloat(data[4].(string), 64)
+		side, err := order.StringToOrderSide(resp.Result[x].Side)
 		if err != nil {
 			return err
 		}
-		volume, err := strconv.ParseFloat(data[5].(string), 64)
-		if err != nil {
-			return err
+		fills[x] = fill.Data{
+			Timestamp:    time.UnixMilli(int64(resp.Result[x].CreateTimeMs)),
+			Exchange:     g.Name,
+			CurrencyPair: currencyPair,
+			Side:         side,
+			OrderID:      resp.Result[x].OrderID,
+			TradeID:      strconv.FormatInt(resp.Result[x].ID, 10),
+			Price:        resp.Result[x].Price,
+			Amount:       resp.Result[x].Amount,
 		}
+	}
+	return g.Websocket.Fills.Update(fills...)
+}
 
-		p, err := currency.NewPairFromString(data[7].(string))
-		if err != nil {
-			return err
+func (g *Gateio) processSpotBalances(data []byte) error {
+	type response struct {
+		Time    int64           `json:"time"`
+		Channel string          `json:"channel"`
+		Event   string          `json:"event"`
+		Result  []WsSpotBalance `json:"result"`
+	}
+	var resp response
+	err := json.Unmarshal(data, &resp)
+	if err != nil {
+		return err
+	}
+	for x := range resp.Result {
+		code := currency.NewCode(resp.Result[x].Currency)
+		g.Websocket.DataHandler <- account.Change{
```

Comment:
suggestion: consider sending slice instead and updating engine and for all slices.

Reply:
bump